### ISSUE TITLE
Experiments with throughput / performance of write operations

### ISSUE THREAD

**Issue by Maze112** [aritchie/bluetoothle#343]
# IF YOU DON'T ANSWER THIS TEMPLATE - THE BOT WILL AUTOMATICALLY CLOSE YOUR ISSUE!

## Please check all of the platforms you are having the issue on (if platform is not listed, it is not supported)

 - [ ] iOS
 - [x] Android
 - [ ] UWP
 - [ ] .NET Standard

## Feature Library

 - [ ] Core
 - [x] BluetoothLE
 - [ ] Locations
 - [ ] HTTP Transfers
 - [ ] Notifications
 - [ ] Sensors

## Version of Library
6.1.3 (situation should be the same up to 6.2.6, tough)

## Version of OS(s) listed above with issue
Android 6.0, 7.0, 9.0 with HW: Sony Experia, ZTE Blade A512, couple of Samsung devices, Nokia (model not known at the moment)

Hi there,
thanks for your work and the library so far! : - )

This is neither an actual bug report nor a real feature request. I wanted to ask about and compare throughput values achieved by the current state of the BluetoothLE library. I wanted to find ideal configurations for the connection.

In short: What are real-world throughput values for write operations? Due to the way the bluetooth stacks operate the measured values ideally transfer a larger amount of data to another device. For example: 100kB of data split into 20b packets. The measured times for a single write operation might be interesting as well despite being somewhat difficult to interpret.

I expected values around 2.5kB / s which is already very conservative compared to various iOS guides, Android forums and Nordic chip forums.

Using defaults for PauseBetweenInvocations (100ms) and write WITH response: ~0.125kB / s
Setting PauseBetweenInvocations to 0ms and write WITH response: ~0.4kB / s
Using defaults for PauseBetweenInvocations (100ms) and write WITHOUT response: 0.2kB / s
Setting PauseBetweenInvocations to 10ms and write WITHOUT response: ~1.8kB / s but with the high risk of overflowing bluetooth stack buffers internally and risking crashes on different platforms.
Setting PauseBetweenInvocations to 0ms and write WITHOUT response: almost guaranteed exceptions during writing to the characteristic

Note that I'm in control of the FW of the other bluetooth device so I can check parameters such as the interval (which is very small at only 9 * 1.25ms = 11.25ms). It seems that there is no real room for optimization of the parameters from the Android side so far. Setting the ConnectionPriority for example does not "help" on my test devices since the interval is already very short (on some devices even at 7.5ms which is the shortest possible according to the specs). Even when we account for only 4 MTUs within one interval we would get a theoretical value of 4 * 80b / 7.5ms => ~11kB / s

I would like to see a setup using BluetoothLE that achieves higher throughput than I did. Maybe there is still something wrong here?

Kind regards,
Mike

## Steps to Reproduce
Since it is about throughput and performance: Every example that measures the data transfer rates should be suitable

## Expected Behavior
Throughput of roughly 2.5kB / s which should be very conservative compared to various calculations found in Nordic forums, for example.

## Actual Behavior
The throughput is way below the expected values. The details are given in the issue description above.

**Comment by aritchie** [comment]
Great tests and I appreciate the numbers.  I'm afraid I don't have anything in terms of numbers to compare against, but I can give you a few of hints.

1) You should always do write with response.  I know it slows things down, but writes do tend to fail.
2) If you use one of the WriteBlob functions, it will use transactions and tight loop.  It can help a bit
3) On Android, you should run RequestMtu to a max rate of 512 (you will never get this value, but it should be higher than the default 20 bytes).  iOS will auto-negotiate this value.

That all being said, if you need any sort of performance - BLE is not your guy.  You are better off going classic Bluetooth.  


I usually don't use BLE if I need any sort of performance.  I tend

**Comment by Maze112** [comment]
Thanks for the response. Yes BLE is definitely not designed for high throughput. Nevertheless, tasks such as updating the firmware of a low energy device need to achieve data rates that are sufficient for the user. Talking about 0.1 to 0.2kB / s is simply not enough even for small images of 150kB.

Choosing an MTU that is larger 20 byte is not recommended right now since there are tons of devices out there that do not allow this feature yet. This, of course, will change in a few years from now. In general you are completely right that the MTU should be negotiated between the devices but for now I focus my experiments on the generation of phones between 2015 and 2017.

My following thoughts are closely related to [PauseBetweenInvocations, what does it means? #252](https://github.com/aritchie/bluetoothle/issues/252)

Writing larger data WITH response slows down the process so much that it's almost unusable even for smaller firmware updates. Using writes WITHOUT response should work on Android as well as on iOS from what I've seen so far.

**Android:** The usual callback (that is called when a response is received after a write) should also be called when writing WITHOUT responses. It usually simply means that the buffer inside the stack has enough room again for some data. Using this callback mechanism should optimally utilize the data rate as discussed [in this stackoverflow post](https://stackoverflow.com/questions/43741849/oncharacteristicwrite-and-onnotificationsent-are-being-called-too-fast-how-to/43744888#43744888). I think that there might be Android versions in combination with HW devices that still require the [write WITHOUT followed by write WITH response hack](https://stackoverflow.com/a/44048096/2104425) as in iOS (see below).

**iOS:** On iPhones it seems that the callback is not called correctly though. There should be 2 approaches to fix this. The one is a [check for a flag that tells if data can be transmitted again](https://stackoverflow.com/questions/33442472/ios-bluetooth-le-proper-way-to-write-without-response/52527853#52527853). Due to a lack of documentation this might not always work or might be tricky to implement in a stable fashion. So the second approach is to write WITHOUT responses for a few times followed by a write WITH response as discussed above for Android.

**What I have right now with the current state of the library:**
Note that the following measurements were done on a ZTE A512 with Android 6.0 which chooses 9 * 1.25 = 11.25ms for the interval during my transmission tests which is quite a good value compared to my other test devices (Sony Xperia chooses 40 * 1.25 = 50ms which sucks!). 
I use 10ms for PauseBetweenInvocations. Whenever I send data packages I do it by using writes WITH response to be safe in the beginning. In case that larger data such as firmware images (more than 1 MTU) has to be sent, all consecutive writes of the same data blob are written as 3 writes WITHOUT and every fourth WITH response. This way I achieved a stable transmission with roughly 0.5kB / s. Note that 1 vs. 3 was chosen since 4 packets per interval are very common for phones. Empirically speaking this feels quite safe to do but one could increase that to probably 1 vs. 5 (got 0.64 kB / s this way) or something like that. Also note that these ratios do not actually relate to the number of packets according to any documentation! It's just about experiments and best guesses for anyone playing around with it!

**Now my suggestions for the BluetoothLE library:**
- **Android and iOS:** Remove the utilization of the PauseBetweenInvocations for all writes WITH response since they should not be needed at all. This is actually a problem since it slows down the writes unnecessarily.
- **Android:** Implement the recommended callback mechanism for writes WITHOUT response similar to writes WITH response but keep the configurable delay in there for everyone that needs it. Every API user can then implement the swapping between writes WITH and WITHOUT responses as they wish.
- **iOS:** Maybe implement the checking of the flag but this might not be needed when implementing the first two suggestions.

**Comment by aritchie** [comment]
iOS doesn't use the queue or pausing mechanism at all, so it won't be a factor here.

You are correct, I can remove the PauseBetweenInvocations for Write WITH Response, but this requires a bit of work.  The callback for WriteWithoutResponse has not proven to work with a bunch of devices I have.  I'll play around with it again.

Thanks for the great examples and digging.  This is helpful.  I'll see what can be done

**Comment by Maze112** [comment]
Thanks, sounds great! I'll be happy to test any changes you make on my available devices and compare throughput and stability with what I have so far to post results here again : - )

**Comment by aritchie** [comment]
I have been quite active on trying to work this out.  

1) I've restored the iOS functionality for Write without response.  It does get up to 2.5 kbs
2) Android has had its Pause functionality removed without any issues so far
3) Android has a working write without response callback now

iOS did get some positive benefits.  Android on the other hand is still subpar in speed (at least on my device).  I did get a much better result, but still only 1.3 kbs on average which is 1/2 of iOS.  

You can find these efforts here: 
https://github.com/shinyorg/shiny/tree/dev/src/Shiny.BluetoothLE and a decent testing sample here:
https://github.com/shinyorg/shinysamples/tree/master/Samples/BluetoothLE

I'd appreciate if you could run your tests with the new library.  I will backport this functionality to this library if needed and I can get it to decent speeds.

**Comment by Maze112** [comment]
Hey, thanks for the effort! Right now I'm only using BluetoothLE without any Shiny integration so I would need to invest some time there. I will report back here as soon as I have new test results.

**Comment by aritchie** [comment]
You actually don't have to put any investment in.  The samples I linked contain almost everything you need to test with.  Since you have a remote device and different types of android devices, I want to see if things have improved (it also has toggles to flip things off - queuing, main thread, etc).

**Comment by Maze112** [comment]
I wanted to compare more or less the same scenarios after your changes so I decided to integrate Shiny to replace BluetoothLE temporarily. I know that the direct comparision cannot be made anymore due to internal changes but still my measurements are very rough in the first place anyways. I wanted the surroundings to stay the same.

**Stability in Android** (Sony Android 7, Samsung S7 Android 8, ZTE A512 Android 6) : 
- Using writes WITH response worked as stable as before.
- Using writes WITHOUT response worked stable across all my data transfer experiments. Congrats so far!

**Performance in Android** (same devices as above): 
- Using writes WITH response has similar performance to 6.2.6. of BluetoothLE (but see comment below).
- Using writes WITHOUT response has a HUGE performance impact! A 230KB firmware update took roughly between 3:00 and 4:30 with even larger connection intervals chosen by the OSes. From what I saw so far I would say that the data rate is 2 to 4 times faster than before! Congrats again!

**iOS**: Sorry, I did not perform any experiments here.

A note that I wanted to post here a week ago: I upgraded to BluetoothLE 6.2.6 from 6.1.3 and I noticed a performance hit of roughly 15-20%. Unfortunately, I didn't find the time yet to check the GIT history to maybe nail the impact down to a specific change. It might not be that important though if you could backport your fixes from Shiny to BluetoothLE. I would then perform in-depth experiments with all of my devices on Android again. I could also include 3 different "new" iOS devices that have become available to me.

**Comment by aritchie** [comment]
Hi @Maze112 - Sorry for the delayed response.  I wanted to thank you for the excellent testing and great feedback.  It is very rare that I get these kind of posts and I really do appreciate them.  As such, I have made the task of porting to shiny almost minimal at this point if you try out the latest version.  I've documented how to skip most of the dependency injection portion here: https://shinydocs.azurewebsites.net/docs/nodi .  If you still find this problematic, let me know and I'll be happy to a) do more to assist with the porting in terms of how the library is used and/or b) back port to this library.

Also, I haven't given up trying to further improve perf on Android.  The opportunities on how to do this at this point are difficult at best.  

I'm going to close this issue, but feel free to respond when you have a moment

**Comment by Maze112** [comment]
Thanks so far!
I understand that you would like your library users to make the transition to Shiny but the "beta" state is holding me back a bit ; - ) Not sure how I will proceed yet.
Having stable 1.3KB/s in Android should be fine : - )

**Comment by KRA2008** [comment]
> That all being said, if you need any sort of performance - BLE is not your guy. You are better off going classic Bluetooth.

I've been working to integrate this package into my project for a while now and was very optimistic about it, but only recently got to the point where I was able to reliably measure performance. I was really confused by what I was seeing in terms of performance and then found this thread. I realize now that I had been making the assumption that this package was compatible with all modern bluetooth versions/flavors up to and including BLE, I did not realize that it may be exclusively for BLE. This thread makes it seem like my understanding of the different versions and flavors of bluetooth that exist may have been incorrect. Is it true that this package only works for the low energy flavor of bluetooth and as such is limited to these low performance rates and CANNOT reach the bluetooth "classic" performance rates?
